### PR TITLE
feat(wire.CloudPublisher): Set as body a property removing it from metrics

### PR DESCRIPTION
**Pull request by marcomatteo**
Added "remove.body.from.metrics" boolean property in the component metatype.
If set to True and the "set a property as body" is not empty, remove the property from the KuraPayload metrics and KuraMessage properties.
The default value is false and preserve the current behaviour.

Added the test cases and refactor the CloudPublisherTest class with scenarios.

**Related Issue:** This PR fixes the issue #4292

### CODE REVIEW

File: kura/org.eclipse.kura.wire.component.provider/OSGI-INF/metatype/org.eclipse.kura.wire.CloudPublisher.xml [hunk 1]
```diff
@@ -47,6 +47,15 @@
...
             description="If set to a non empty value, the publisher will set the message body to the value of the provided STRING or BYTE_ARRAY metric.">
         </AD>
+        
+        <AD id="remove.body.from.metrics"
+            name="Remove body envelope property from metrics"
+            type="Boolean"
+            cardinality="0"
+            required="true"
+            default="false"
+            description="If Set body from envelope property is not an empty value, the publisher will remove from the message metrics the property added in the message body.">
```
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.component.provider/OSGI-INF/metatype/org.eclipse.kura.wire.CloudPublisher.xml]: Maybe rephrase that, the description it is not so clear to me.

Something like: Set true to remove the metric from the message that is set in the message body.

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java [hunk 2]
```diff
@@ -81,7 +81,7 @@ public final class CloudPublisher implements WireReceiver, ConfigurableComponent
      * Binds the Wire Helper Service.
      *
      * @param wireHelperService
-     *            the new Wire Helper Service
+     *                              the new Wire Helper Service
```
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java]: What is happening here? Use the kura formatter and cleanup profiles if using Eclipse (https://github.com/eclipse/kura/tree/develop/kura/setups/formatting). Replace tabs with spaces.

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java [hunk 3]
```diff
@@ -187,14 +187,14 @@ public void updated(final Wire wire, final Object value) {
...
+     *                       the {@link WireRecord}
      * @return the Kura payload
      * @throws NullPointerException
-     *             if the {@link WireRecord} provided is null
+     *                                  if the {@link WireRecord} provided is null
      */
     private KuraPayload buildKuraPayload(final WireRecord wireRecord) {
         requireNonNull(wireRecord, "Wire Record cannot be null");
-        final KuraPayload kuraPayload = new KuraPayload();
+        KuraPayload kuraPayload = new KuraPayload();
```
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java]: Why did you remove the `final` keyword?

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisherOptions.java [hunk 4]
```diff
@@ -82,4 +83,9 @@ Optional<String> getBodyProperty() {
 
         return Optional.of(property);
     }
+
+    Boolean getRemoveBodyPropertyFromMetrics() {
```
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisherOptions.java]: Use the primitive type here, it eases also the usage later on.

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisherOptions.java [hunk 5]
```diff
@@ -82,4 +83,9 @@ Optional<String> getBodyProperty() {
...
+        final Object propertyRaw = this.properties.get(CONF_REMOVE_BODY_PROPERTY);
```
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisherOptions.java]: To be on the safe side maybe it is better to do something like:
```java
return (boolean) this.properties.getOrDefault(CONF_REMOVE_BODY_PROPERTY, false);
```

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java [hunk 6]
```diff
@@ -280,9 +285,17 @@ private void publish(final List<WireRecord> wireRecords) {
 
     private Map<String, Object> buildKuraMessageProperties(final WireRecord wireRecord) {
         Map<String, TypedValue<?>> wireRecordProps = wireRecord.getProperties();
+        Set<String> metricProperties = new HashSet<>(wireRecordProps.keySet());
```
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java]: Can you refactor this method? It is quite cryptic. Maybe add more meaningful names to variables.

**Review by marcellorinaldo — CHANGES_REQUESTED** [review]

**Comment by marcellorinaldo** [comment]
Thanks for the PR @marcomatteo. When you have the time, could you please address the review?

**Comment by marcomatteo** [comment]
Thank you @marcellorinaldo for the review. I'll answer and fix soon.

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java (hunk 2 above)
Comment (marcomatteo) [reply kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java]: You're right, I followed this [documentation page](https://wiki.eclipse.org/Kura/Getting_Started) and was the cleanup/formatter v2 and not the updated v3.

File: kura/org.eclipse.kura.wire.component.provider/OSGI-INF/metatype/org.eclipse.kura.wire.CloudPublisher.xml (hunk 1 above)
Comment (marcomatteo) [reply kura/org.eclipse.kura.wire.component.provider/OSGI-INF/metatype/org.eclipse.kura.wire.CloudPublisher.xml]: Changed in something more simple: "Set true to remove the metric that is sent as body of the message."

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java (hunk 3 above)
Comment (marcomatteo) [reply kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java]: I want to know it too.

File: kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java (hunk 6 above)
Comment (marcomatteo) [reply kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java]: Updated with the same logic of metrics: first adding all the metrics to the to the Map, then remove only if the flag is true.

**Comment by MMaiero** [comment]
Hi @marcomatteo,
are the changes to this PR done?
Can we ask @marcellorinaldo to perform another pass for the review?

Thanks

**Comment by marcomatteo** [comment]
Hi @MMaiero,
yes, I'm waiting for the review.
Thank you in advance.

**Comment by marcellorinaldo** [comment]
Also here, the Eclipse CI cannot retrieve Sonar status for Kura project.

image

**Review by marcellorinaldo — APPROVED** [review]
